src/providers/wfs/qgswfsprovidermetadata.cpp

```diff
@@ -0,0 +1,384 @@
+  propertyIsEqualToElem.appendChild( functionElem );
+  QDomAttr attrFunctionName = doc.createAttribute( QStringLiteral( "name" ) );
+  attrFunctionName.setValue( function );
+  functionElem.setAttributeNode( attrFunctionName );
+  QDomElement valueReferenceElem = doc.createElement(
+                                     ( caps.version.startsWith( "2.0" ) ) ? filterPrefix + QStringLiteral( ":ValueReference" ) : filterPrefix + QStringLiteral( ":PropertyName" ) );
+  functionElem.appendChild( valueReferenceElem );
+  valueReferenceElem.appendChild( doc.createTextNode( geometryElement ) );
+  QDomElement literalElem = doc.createElement( filterPrefix + QStringLiteral( ":Literal" ) );
+  propertyIsEqualToElem.appendChild( literalElem );
+  literalElem.appendChild( doc.createTextNode( QStringLiteral( "true" ) ) );
+
+  return doc.toString();
+}
+
+QString QgsWFSProvider::buildIsNullGeometryFilter( const QgsWfsCapabilities::Capabilities &caps,
+    const QString &geometryElement )
+{
+  /* Generate something like:
+   <fes:Filter xmlns="http://www.opengis.net/fes/2.0">
+    <fes:PropertyIsNull>
+        <fes:ValueReference>xplan:geltungsbereich</fes:ValueReference>
+    </fes:PropertyIsNull>
+  </fes:Filter>
+  */
+
+  QDomDocument doc;
+  QDomElement filterElem =
+    ( caps.version.startsWith( QStringLiteral( "2.0" ) ) ) ?
+    doc.createElementNS( QStringLiteral( "http://www.opengis.net/fes/2.0" ), QStringLiteral( "fes:Filter" ) ) :
+    doc.createElementNS( QStringLiteral( "http://www.opengis.net/ogc" ), QStringLiteral( "ogc:Filter" ) ) ;
+  doc.appendChild( filterElem );
+  QString filterPrefix( caps.version.startsWith( "2.0" ) ? QStringLiteral( "fes" ) : QStringLiteral( "ogc" ) );
+  QDomElement propertyIsNullElem = doc.createElement( filterPrefix + QStringLiteral( ":PropertyIsNull" ) );
+  filterElem.appendChild( propertyIsNullElem );
+  QDomElement valueReferenceElem = doc.createElement(
+                                     ( caps.version.startsWith( "2.0" ) ) ? filterPrefix + QStringLiteral( ":ValueReference" ) : filterPrefix + QStringLiteral( ":PropertyName" ) );
+  propertyIsNullElem.appendChild( valueReferenceElem );
+  valueReferenceElem.appendChild( doc.createTextNode( geometryElement ) );
+
+  return doc.toString();
+}
+
+QString QgsWFSProvider::buildGeometryCollectionFilter( const QgsWfsCapabilities::Capabilities &caps,
+    const QString &geometryElement )
+{
+  QDomDocument doc;
+  QDomElement filterElem =
+    ( caps.version.startsWith( QStringLiteral( "2.0" ) ) ) ?
+    doc.createElementNS( QStringLiteral( "http://www.opengis.net/fes/2.0" ), QStringLiteral( "fes:Filter" ) ) :
+    doc.createElementNS( QStringLiteral( "http://www.opengis.net/ogc" ), QStringLiteral( "ogc:Filter" ) ) ;
+  doc.appendChild( filterElem );
+  QString filterPrefix( caps.version.startsWith( "2.0" ) ? QStringLiteral( "fes" ) : QStringLiteral( "ogc" ) );
+
+  QDomElement andElem = doc.createElement( filterPrefix + QStringLiteral( ":And" ) );
```

Comment:
This `And` element includes only the `!PropertyIsNull` filter but not the `!IsPoint`, `!IsCurve`, `!IsSurface` filters, resulting in an [invalid filter](https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs?SERVICE=WFS&REQUEST=GetFeature&VERSION=2.0.0&TYPENAMES=xplan:BP_AnpflanzungBindungErhaltung&COUNT=1&SRSNAME=EPSG:25832&FILTER=%3Cfes:Filter%20xmlns:fes%3D%22http://www.opengis.net/fes/2.0%22%3E%0A%20%3Cfes:And%3E%0A%20%3Cfes:Not%3E%0A%20%3Cfes:PropertyIsNull%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:PropertyIsNull%3E%0A%20%3C/fes:Not%3E%0A%20%3C/fes:And%3E%0A%20%3Cfes:PropertyIsEqualTo%3E%0A%20%3Cfes:Function%20name%3D%22IsPoint%22%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:Function%3E%0A%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%3C/fes:PropertyIsEqualTo%3E%0A%20%3Cfes:PropertyIsEqualTo%3E%0A%20%3Cfes:Function%20name%3D%22IsCurve%22%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:Function%3E%0A%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%3C/fes:PropertyIsEqualTo%3E%0A%20%3Cfes:PropertyIsEqualTo%3E%0A%20%3Cfes:Function%20name%3D%22IsSurface%22%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:Function%3E%0A%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%3C/fes:PropertyIsEqualTo%3E%0A%3C/fes:Filter%3E%0A).

Reply:
good catch. sorry for that. fixed per e1c660dd1c
Now manual testing of ``vl = QgsVectorLayer("pagingEnabled='true' preferCoordinatesForWfsT11='false' restrictToRequestBBOX='1' srsname='EPSG:25832' typename='xplan:BP_AnpflanzungBindungErhaltung' url='https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs' version='auto' geometryTypeFilter='GeometryCollection'", 'test', 'WFS')`` results in the correct https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs?SERVICE=WFS&REQUEST=GetFeature&VERSION=2.0.0&TYPENAMES=xplan:BP_AnpflanzungBindungErhaltung&COUNT=1&SRSNAME=EPSG:25832&FILTER=%3Cfes:Filter%20xmlns:fes%3D%22http://www.opengis.net/fes/2.0%22%3E%0A%20%3Cfes:And%3E%0A%20%20%3Cfes:Not%3E%0A%20%20%20%3Cfes:PropertyIsNull%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:PropertyIsNull%3E%0A%20%20%3C/fes:Not%3E%0A%20%20%3Cfes:PropertyIsEqualTo%3E%0A%20%20%20%3Cfes:Function%20name%3D%22IsPoint%22%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:Function%3E%0A%20%20%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%20%3C/fes:PropertyIsEqualTo%3E%0A%20%20%3Cfes:PropertyIsEqualTo%3E%0A%20%20%20%3Cfes:Function%20name%3D%22IsCurve%22%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:Function%3E%0A%20%20%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%20%3C/fes:PropertyIsEqualTo%3E%0A%20%20%3Cfes:PropertyIsEqualTo%3E%0A%20%20%20%3Cfes:Function%20name%3D%22IsSurface%22%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:Function%3E%0A%20%20%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%20%3C/fes:PropertyIsEqualTo%3E%0A%20%3C/fes:And%3E%0A%3C/fes:Filter%3E%0A request being emitted

Reply:
Thanks and no worries Even, I'm happy to give some useful feedback and still impressed what you've created here.